src/bin/wasmtime.rs

```diff
@@ -99,11 +99,176 @@ impl Wasmtime {
+        let old = try_parse_old();
+        match (new, old) {
+            // Both parsers succeeded. This means that it's likely that no
+            // options to configure execution, e.g. `--disable-logging`, were
+            // used in the old parser or the new.
+            //
+            // The result of parsing can still differ though. For example:
+            //
+            //      wasmtime foo.wasm --invoke foo
+            //
+            // would previously be parsed as passing `--invoke` to Wasmtime but
+            // the new parser instead parses that as passing the flag to
+            // `foo.wasm`.
+            //
+            // In this situation use the `DEFAULT_OLD_BEHAVIOR` constant to
+            // dictate which wins and additionally print a warning message.
+            (Ok(new), Ok(old)) => {
+                if new == old {
+                    return new.execute();
+                }
+                if DEFAULT_OLD_BEHAVIOR {
+                    eprintln!(
+                        "\
+warning: this CLI invocation of Wasmtime will be parsed differently in future
+         Wasmtime versions, see this online issue for more information:
+         https://github.com/bytecodealliance/wasmtime/issues/7384
+         you can use `WASMTIME_NEW_CLI=0` to temporarily disable this warning\
+"
+                    );
+                    old.execute()
+                } else {
+                    eprintln!(
+                        "\
+warning: this CLI invocation of Wasmtime is parsed differently now than it was
+         in Wasmtime 13 and prior, and the new semantics are going to be used
+         instead of the old semantics -- see this issue for more information:
+         https://github.com/bytecodealliance/wasmtime/issues/7384
+         you can use `WASMTIME_NEW_CLI=1` to disable this warning\
+"
+                    );
+                    new.execute()
+                }
+            }
+
+            // Here the new parser succeeded where the old one failed. This
+            // could indicate for example that new options are being passed on
+            // the CLI.
+            //
+            // In this situation assume that the new parser is what's intended
+            // so execute those semantics.
+            (Ok(new), Err(_old)) => new.execute(),
+
+            // Here the new parser failed and the old parser succeeded. This
+            // could indicate for example passing old CLI flags.
+            //
+            // Here assume that the old semantics are desired but emit a warning
+            // indicating that this will change in the future.
+            (Err(_new), Ok(old)) => {
+                eprintln!(
+                    "\
+warning: this CLI invocation of Wasmtime is going to break in the future, for
+         more information see this issue online:
+         https://github.com/bytecodealliance/wasmtime/issues/7384
+         use `WASMTIME_NEW_CLI=0` to temporarily disable this warning\
```

Comment:
Likewise here; perhaps we can have a specific message ("this invocation is going to break" vs. "this invocation is parsed differently by old and new") and then a general informational blob afterward?